server/config/index.js

```diff
@@ -14,6 +14,29 @@ var config = {
     password: process.env.POSTGRES_PASSWORD || 'postgres',
     host: process.env.POSTGRES_HOST || '127.0.0.1',
     port: process.env.POSTGRES_PORT || '5432'
+  },
+  schemaListenerConfig: undefined
+}
+
+if (process.env.SCHEMA_LISTENER_CONFIG) {
```

Comment:
I'm not a big fan of supplying a JSON string as an environment variable. I've had previous (ops) experience where this was the most difficult thing ever.
technologies like Jenkins, CircleCI, Kubernetes, docker-compose, bash etc. All of these handle and escape these strings slightly differently and in my experience, It's really hard to get this to work everywhere.

I know it seems hacky but base64 is super portable. If really you want to supply this config as a JSON string I would suggest we assume this to be base64 encoded and then decode it here as part of startup. It would look something like this:

```
Buffer.from(process.env.SCHEMA_LISTENER_CONFIG, 'base64').toString()
```

About the JSON parsing... I had a discussion about user supplied config when working on aerogear-app-metrics. It was to do with something trivial like setting log levels via environment variables. I originally wanted to start the app with reasonable defaults in a case where the user supplied broken config.

The conclusion we arrived at was if a user supplied broken configuration, then the best option is to just crash the service because that service might start behaving in a way the user didn't expect and they won't know why.

Reply:
About JSON in env vars: you're right that it could cause problems. We need to change it.
I think base64 idea is good. But, one thing we need to keep in mind is, how easy it is to create a bas64 encoded string with those techs (Jenkins, CircleCI, Ansible etc). Let's think about this for a minute.

The other way, and crappier way, is to use multiple env vars like:
For pg node Postgres channels:
- SCHEMA_LISTENER_TYPE: 'postgres'
- POSTGRES_CHANNEL_NAME: 'aerogear-something'
For the imaginary message broker XYZ:
- SCHEMA_LISTENER_TYPE: 'xyz'
- XYZ_TOPIC_NAME: 'aerogear-something'
- XYZ_HOST: 'example.com'
- XYZ_USERNAME: 'user'
- ...

The reason I went with JSON config in env vars is I didn't like all those strings (env var keys) that are used in some cases and not used in other cases.
I just wanted to bring this way to discussion so that we can have answers when somebody complains about base64 :smiling_imp:

Reply:
> The conclusion we arrived at was if a user supplied broken configuration, then the best option is to just crash the service because that service might start behaving in a way the user didn't expect and they won't know why.

This is a good point. Then I suggest that we get rid of all the defaults in the code and just have those values in the dev scripts. Then we can complain whenever something is missing. Purpose of defaults to me are to start things faster on devl anyway.

Reply:
Yup no config is fine, it usually means we're in dev or that the user is happy with the defaults. Broken config means the user messed up and we shouldn't start in an unknown state.

With regards to the question of how difficult is base64 encoding something? It's pretty simple anywhere with a bash shell.

```bash
echo 'Hello world' | base64
```

Reply:
I'm not a massive fan of the base64 thing but I've genuinely spent an entire week trying to escape a JSON string so a CI server could feed it into an app before and I don't ever want to do that again 😄 

I guess the reason you're not a fan of the other solution is because it's inherently tied to Postgres and it kinda breaks the abstraction layer we're trying to place above that.

To be honest, as long as the underlying code is abstracted well enough that we could swap it out in the future, I wouldn't mind the other approach because refactoring a config file is a minor thing.

I will defer this to you :)